crates/bin/prove_block/src/main.rs

```diff
@@ -0,0 +1,741 @@
+use starknet_os::crypto::poseidon::PoseidonHash;
+use starknet_os::error::SnOsError::Runner;
+use starknet_os::execution::helper::{ContractStorageMap, ExecutionHelperWrapper};
+use starknet_os::io::input::StarknetOsInput;
+use starknet_os::starknet::business_logic::fact_state::contract_class_objects::{
+    get_ffc_for_contract_class_facts, ContractClassLeaf,
+};
+use starknet_os::starknet::business_logic::fact_state::contract_state_objects::ContractState;
+use starknet_os::starknet::business_logic::fact_state::state::SharedState;
+use starknet_os::starknet::business_logic::utils::write_class_facts;
+use starknet_os::starknet::starknet_storage::{CommitmentInfo, OsSingleStarknetStorage};
+use starknet_os::starkware_utils::commitment_tree::base_types::{Height, Length, NodePath, TreeIndex};
+use starknet_os::starkware_utils::commitment_tree::binary_fact_tree::BinaryFactTree;
+use starknet_os::starkware_utils::commitment_tree::patricia_tree::nodes::{BinaryNodeFact, EdgeNodeFact};
+use starknet_os::storage::cached_storage::CachedStorage;
+use starknet_os::storage::storage::{Fact, FactFetchingContext, Storage, StorageError};
+use starknet_os::utils::felt_vm2api;
+use starknet_os::{config, run_os};
+use starknet_os_types::contract_class::GenericCasmContractClass;
+use starknet_types_core::felt::Felt;
+use types::starknet_rs_to_blockifier;
+
+use crate::types::starknet_rs_tx_to_internal_tx;
+
+mod reexecute;
+mod types;
+
+#[derive(Parser, Debug)]
+struct Args {
+    /// Block to prove.
+    #[arg(long = "block-number")]
+    block_number: u64,
+
+    /// RPC endpoint to use for fact fetching
+    #[arg(long = "rpc-provider", default_value = "http://localhost:9545")]
+    rpc_provider: String,
+}
+
+fn jsonrpc_request(method: &str, params: serde_json::Value) -> serde_json::Value {
+    json!({
+        "jsonrpc": "2.0",
+        "id": "0",
+        "method": method,
+        "params": params,
+    })
+}
+
+async fn post_jsonrpc_request<T: DeserializeOwned>(
+    client: &reqwest::Client,
+    rpc_provider: &str,
+    method: &str,
+    params: serde_json::Value,
+) -> Result<T, reqwest::Error> {
+    let request = jsonrpc_request(method, params);
+    let response = client.post(format!("{}/rpc/v0_7", rpc_provider)).json(&request).send().await?;
+
+    #[derive(Deserialize)]
+    struct TransactionReceiptResponse<T> {
+        result: T,
+    }
+
+    println!("Response status: {}", response.status());
+    let response_text = response.text().await?;
+    let response: TransactionReceiptResponse<T> =
+        serde_json::from_str(&response_text).unwrap_or_else(|_| panic!("Error: {}", response_text));
+    Ok(response.result)
+}
+
+#[derive(Clone, Deserialize)]
+struct EdgePath {
+    len: u64,
+    value: Felt252,
+}
+
+#[derive(Clone, Deserialize)]
+enum TrieNode {
+    #[serde(rename = "binary")]
+    Binary { left: Felt252, right: Felt252 },
+    #[serde(rename = "edge")]
+    Edge { child: Felt252, path: EdgePath },
+}
```

Comment:
I think it's time to split this file into several modules. All the Pathfinder-specific types should go in a dedicated module for example.